src/Npgsql/ConnectorPool.cs

```diff
@@ -235,37 +227,46 @@ internal async ValueTask<NpgsqlConnector> AllocateLong(NpgsqlConnection conn, Np
                                 newPruningTimer.Dispose();
                             }
                         }
+
+                        Counters.NumberOfPooledConnections.Increment();
+                        Counters.NumberOfActiveConnections.Increment();
+                        CheckInvariants(State);
+                        return connector;
                     }
                     catch
                     {
-                        // Physical open failed, decrement the open counter back down.
+                        // Physical open failed, decrement the open and busy counter back down.
                         conn.Connector = null;
                         Interlocked.Decrement(ref State.Open);
                         throw;
                     }
-
-                    Counters.NumberOfPooledConnections.Increment();
-                    Counters.NumberOfActiveConnections.Increment();
-                    Interlocked.Increment(ref State.Busy);
-                    CheckInvariants(State);
-                    return connector;
                 }
 
-                if (Volatile.Read(ref State.Busy) == _max)
+                // We may be racing a connector Close, either it decrements Open and Idle state between our reads
+                // or we read right after both decrements, in any case causing busy to be less than _max.
+                // When we read in between the decrements, idleCount will be > 0 and our next check in this loop will
+                // call TryAllocateFast, 'incorrect' but inconsequential and corrects once the Idle decrement happens.
+                // When we read right after all decrements, idleCount == 0 and openCount < _max so we loop again
+                // to open a new connector, if we lose that race we'll move into the waiter queue anyway.
+                // Keep in mind, busy is only really read after all other options, when the pool is exhausted.
+                // Pool exhaustion and idle connector closes together is a strange combination, and recollect that
+                // during Close any connectors that weren't idle don't decrement idle, obviating any race.
+                openCount = Volatile.Read(ref State.Open);
+                var idleCount = Volatile.Read(ref State.Idle);
```

Comment:
They are `int`s and all modern CPUs are 64 bit, so make sense to read these values as a single operation.

Reply:
Ah yes I actually janked out my false sharing padding around Open and Idle right before I committed. I concluded in my commit message there's only really one infrequent scenario where that shows up.

Though if we'd read them as one that doesn't really remove the race as they are modified separately as well. Could do a CAS loop there but that feels like unnecessary work for a benign race?

Reply:
(That false sharing padding would make the struct > 64bit)

Commit:
>There's some false sharing during concurrent connector open and release/allocate but during steady state fastallocate/release that's not relevant at all. Not going to add padding for it.

Reply:
I'll leave it like this for now

Reply:
Yeah it's not a lot of work, I'll put it in

Reply:
Hmm I remember now, it gets ugly quickly with the `if (wasIdle)` check guarding the second XADD. So there's only one code path really doing two increments. Also doing that atomic cas would have to re-introduce the FieldOffsets and an All field, which would precluding any changes to the structure or false sharing padding forever. Changes to the structure of the pool state would require a re-examination of every possible situation in the pool because XADD doesn't exist for short, so these two ints are the max for what state could contain.

I'd really like to keep the current design in here if at all possible, it actually makes the implementation more open to change, and much more flexible. I think the value above your suggestion is that working on any part of the pool would be less of an immediate full rework and think through.

Reply:
> it gets ugly quickly with the if (wasIdle) check guarding the second XADD. So there's only one code path really doing two increments.

Right, so Interlocked.Decrement(open) in one branch, and a CAS loop in another, no? Is that a big deal?

> Also doing that atomic cas would have to re-introduce the FieldOffsets and an All field, which would precluding any changes to the structure or false sharing padding forever.

Are you saying you there that there are possible gains in putting padding between the Open and Idle counters? That seems a bit counterintuitive, what exactly do you have in mind?

> Changes to the structure of the pool state would require a re-examination of every possible situation in the pool because XADD doesn't exist for short, so these two ints are the max for what state could contain

Sure, but I don't think we need to reason with such "future-perfect" constraints in mind - the implementation is what it is today... It's very hard to anticipate what design possibilities may arise in the future and what constraints they would impose.

> I'd really like to keep the current design in here if at all possible, it actually makes the implementation more open to change, and much more flexible. I think the value above your suggestion is that working on any part of the pool would be less of an immediate full rework and think through.

I'm not sure I agree - maybe you have a future direction in mind? All we're discussing is defining an additional long encompassing the existing two ints, and using that to read and write atomically for stronger consistency. If a need really arises in the future, we can always relax this and go back to your current design, no?

Reply:
BTW see also my comment below: https://github.com/npgsql/npgsql/pull/2519/files#r305823392

Reply:
>Are you saying you there that there are possible gains in putting padding between the Open and Idle counters? That seems a bit counterintuitive, what exactly do you have in mind?

From this thread https://github.com/npgsql/npgsql/pull/2519#discussion_r304331672

>Sure, but I don't think we need to reason with such "future-perfect" constraints in mind - the implementation is what it is today... It's very hard to anticipate what design possibilities may arise in the future and what constraints they would impose.

Open for extension is something I put a lot of value on, in small companies that often makes your job a whole lot more effective if you stay around for a while. I believe npgsql to be similar in the undermanned aspect.

You're asking to put constraints in that are not strictly necessary, that's fine and not without merit, as it lessens the burden on code comprehension. What irks me is that it increases (almost any) future change thinking work by a lot, as costly knowledge is, 'lost' c.q. invisible, that something more flexible is actually entirely fine.

What about this, I'll push Busy back into the struct as an atomic read of both values and I'll leave a modified comment stating what's possible as well. I'm more than ok with that.

Reply:
Re false sharing, I read your comment above and I agree that there's at least a theoretical potential for helping perf by putting some space between Idle and Open. Is this something you want to try and tackle in this PR, or afterwards?

> What about this, I'll push Busy back into the struct as an atomic read of both values and I'll leave a modified comment stating what's possible as well. I'm more than ok with that.

That's great - I'd also really prefer it if you did the same on the writing side, though. But all that would mean we're dropping the false sharing.

Now for some theoretical nitpicking that may not end up relevant :)

First, from an extensibility/future-proofing point of view, we could look the situation it in the opposite way. While with the *current* implementation, not doing things atomically may be OK, it's  conceivable that we do a change in the future that would invalidate that. At that point we're relying on our memory, reasoning (and possibly on you being there and reviewing :)). In other words, not putting atomicity in now could cause a severe and extremely subtle bug later down the road as changes are made and we don't have the full reasoning (and it's quite complex one). So in my mind, adding this constraint is actually a way of making the pool safer in the future.

Second, if we do find that we need to do things non-atomically (e.g. to prevent false sharing), we can always change the implementation *at that point*. This isn't an API - just an implementation detail - and I prefer not to do things prematurely (especially if they carry risk/complexity as in this case).

In general I'd probably be more on your side, but this lock-free scenario is a bit special in terms of the complexity and risk involved.

Reply:
Agreed, it's entirely fine like this. mostly it'd be a shame to forget how far you could go, if you need it, so this is a good compromise. It's actually also implemented on the write side, or did I miss something?

I don't think it's all too important to tackle the false sharing, it only happens until all connections are opened so perf settle at some point

Reply:
> mostly it'd be a shame to forget how far you could go

I understand that point - you did document it quite extensively in PoolState so hopefully that won't be forgotten.

> It's actually also implemented on the write side, or did I miss something?

Nope, all good :) I'm just reading through the comments, didn't get to Close() yet and thought you only did read.

> I don't think it's all too important to tackle the false sharing, it only happens until all connections are opened so perf settle at some point

Just saw your comment on 68e4a838803a76e24075edeae8e4ab321304a939 and I completely agree. Your change from tracking idle/busy to tracking idle/open means that in normal use there's only one counter that's actually perf-sensitive (idle) - and that's great.

Reply:
Thanks for understanding about the atomicity :)